### PR TITLE
add connector

### LINKED ISSUE #4754: [Site Request] TukangKomik
**Name of the website**
TukangKomik

**Website urls (examples below)**
- Site: https://tukangkomik.com
- Manga List: https://tukangkomik.com/manga/
- Manga example: https://tukangkomik.com/manga/im-an-evil-god/
- Chapter online viewer example: https://tukangkomik.com/im-an-evil-god-chapter-265/

**Languages**
Indonesia

**Website relationship*
-

**Additional details**
Themesia Theme

### PR DESCRIPTION
**Pull request by akn19**
- hinapyon
- tukangkomik ( Fixes https://github.com/manga-download/hakuneko/issues/4754 )
- lianscan
- duniakomik (gabutscan)
- mangaindoMe

### CODE REVIEW

**Comment by MikeZeDev** [comment]
I see nothing wrong but i am fairly new here and i dont know the policy about renamed connectors.

**Comment by akn19** [comment]
> I see nothing wrong but i am fairly new here and i dont know the policy about renamed connectors.

[DISCORD] on 2022/02/10 gabutscan announced moving their website to a new domain for the main reason of optimization and leaving their old website.

it would be nice if the owner of this project added a contributing guide like [vscode](https://github.com/microsoft/vscode/blob/main/CONTRIBUTING.md) and many other open source projects.

sorry if my english is bad.

**Comment by Sheepux** [comment]
you mean this ? https://github.com/manga-download/hakuneko/wiki

**Comment by MikeZeDev** [comment]
So, basically when connector name or group name change its a bad idea to rename a file or change the id.

* You modify the existent connector and keep the id inside (so people dont lose cached manga lists and bookmarks). 
* Or you make a brand new one.

In this PR , classname is still "GabutScans" and does not match the filename btw. Make sure the class name matches the filename.

**Comment by akn19** [comment]
> * You modify the existent connector and keep the id inside (so people dont lose cached manga lists and bookmarks).
> * Or you make a brand new one.
> 
> In this PR , classname is still "GabutScans" and does not match the filename btw. Make sure the class name matches the filename.

done

File: src/web/mjs/connectors/DoujinDesu.mjs [hunk 1]
```diff
@@ -7,14 +7,15 @@ export default class DoujinDesu extends WordPressMangastream {
...
+        this.url = 'https://212.32.226.234';
         this.path = '/manga/page/%PAGE%/';
+        this.requestOptions.headers.set('x-referer', this.url);
 
-        this.querMangaTitleFromURI = 'main#archive div.wrapper h1.title';
-        this.queryMangas = '#main .entries .entry > a';
+        this.queryMangas = 'div.entries article.entry a';
         this.queryChapters = 'div#chapter_list div.epsleft span.lchx a';
+        this.queryPages = 'div.main div img[src]:not([src=""])';
         this.queryChaptersTitle = undefined;
```
Comment (MikeZeDev) [inline src/web/mjs/connectors/DoujinDesu.mjs:17]: If this variable is not used it should be deleted.

Comment (MikeZeDev) [inline src/web/mjs/connectors/DoujinDesu.mjs:16]: Maybe you should delete queryChaptersTitle if its not used?

**Comment by MikeZeDev** [comment]
I tested DoujinDesu manga/ecchi-na-fuushuu-ga-aru-kaso-shuuraku-no-ohanashi/ and i got some error 403 on pictures. Host need https://doujindesu.xxx as referer. 

To add referer for pictures you need to wrap images urls with this.createConnectorURI(pictureurl), override handleConnectorURI and using it to pass the good referer to the picture request
```
    async _handleConnectorURI(payload) {
        let request = new Request(payload, this.requestOptions);
        request.headers.set('x-referer', this.url);
        let response = await fetch(request);
        let data = await response.blob();
        data = await this._blobToBuffer(data);
        this._applyRealMime(data);
        return data;
    }
```

File: src/web/mjs/connectors/LianScans.mjs [hunk 2]
```diff
@@ -8,13 +8,33 @@ export default class LianScans extends WordPressMangastream {
...
+            let mangas = await this._getMangasFromPage(page);
+            mangas.length > 0 ? mangaList.push(...mangas) : run = false;
+        }
+        return mangaList;
+    }
+
+    async _getMangasFromPage(page) {
+        const uri = new URL('/manga/?page=' + page, this.url);
+        const request = new Request(uri, this.requestOptions);
+        let data = await this.fetchDOM(request, 'div.listupd div.bs div.bsx > a');
```
Comment (MikeZeDev) [inline src/web/mjs/connectors/LianScans.mjs:32]: Why not using this.queryMangas which you defined?

**Review by MikeZeDev — CHANGES_REQUESTED** [review]

File: src/web/mjs/connectors/DoujinDesu.mjs (hunk 1 above)
Comment (akn19) [reply src/web/mjs/connectors/DoujinDesu.mjs:16]: > Maybe you should delete queryChaptersTitle if its not used?

chapter list is empty if i delete it

Comment (MikeZeDev) [reply src/web/mjs/connectors/DoujinDesu.mjs:16]: Ok its a template thing i see

File: src/web/mjs/connectors/LianScans.mjs (hunk 2 above)
Comment (MikeZeDev) [inline src/web/mjs/connectors/LianScans.mjs]: there is this to solve. 
        let data = await this.fetchDOM(request, 'div.listupd div.bs div.bsx > a');
could be replaced by  
        let data = await this.fetchDOM(request, this.queryMangas);

Comment (akn19) [reply src/web/mjs/connectors/LianScans.mjs]: Done

**Comment by akn19** [comment]
> I tested DoujinDesu manga/ecchi-na-fuushuu-ga-aru-kaso-shuuraku-no-ohanashi/ and i got some error 403 on pictures. Host need https://doujindesu.xxx as referer.
> 
> To add referer for pictures you need to wrap images urls with this.createConnectorURI(pictureurl), override handleConnectorURI and using it to pass the good referer to the picture request
> 
> ```
>     async _handleConnectorURI(payload) {
>         let request = new Request(payload, this.requestOptions);
>         request.headers.set('x-referer', this.url);
>         let response = await fetch(request);
>         let data = await response.blob();
>         data = await this._blobToBuffer(data);
>         this._applyRealMime(data);
>         return data;
>     }
> ```

this code not work for me

```
:authority: pixdesu.art
:method: GET
:path: /storage/drive/n704ZRlEl35xjVwncoPWDSwTB3SXCN/JoCxHZFEuJQGPwULU8DPczCk4uRl19/2ShcaY7dBV0TmiA5eBCI3DdIPZH5OT/bjtS30eYKpkNH9OTCoBGQFfoyEngEc.webp
:scheme: https
accept: image/webp,image/apng,image/*,*/*
accept-encoding: gzip, deflate, br
accept-language: en-US
sec-ch-ua: 
sec-fetch-dest: empty
sec-fetch-mode: cors
sec-fetch-site: cross-site
user-agent: Mozilla/5.0 (Windows NT 6.1; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/51.32.4805.947 Safari/537.36
```
---
use `this.requestOptions.headers.set('x-referer', this.url);`

```
:authority: pixdesu.art
:method: GET
:path: /storage/drive/n704ZRlEl35xjVwncoPWDSwTB3SXCN/JoCxHZFEuJQGPwULU8DPczCk4uRl19/2ShcaY7dBV0TmiA5eBCI3DdIPZH5OT/25F9Lof6eAvSjIe98UeLyWWR6e9oWG.webp
:scheme: https
accept: image/webp,image/apng,image/*,*/*
accept-encoding: gzip, deflate, br
accept-language: en-US
referer: https://212.32.226.234
sec-ch-ua: 
sec-fetch-dest: empty
sec-fetch-mode: cors
sec-fetch-site: cross-site
user-agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/65.28.1389.132 Safari/537.36
```

**Comment by MikeZeDev** [comment]
Because you did half of the job :).  "HandleConnectorURI' isnt used without a "connectorURI", which i a special way to wrap images links, so hakuneko know there are special things to do, and then use 'HandleConnectorURI'

Like i said : 

>to add referer for pictures you need to wrap images urls with this.createConnectorURI(pictureurl)

Images urls are returned in _getpages. You have to copy getpages from the template WordPressMangastream.mjs, and make is so it returns this.createConnectorURI(pictureurl) array .

Exemple of use : https://github.com/manga-download/hakuneko/pull/5237/files **(dont copy it)**. Check how i modified the getpages from the template so you'll understand.

**Review by MikeZeDev — APPROVED** [review]

**Pull request merged**